add configuration models for additional context

elrayle opened the pull request:
These are the classes to process the context configuration.  They are not being used yet.

Remaining work:
* In search_config, read in and process the context configuration using these classes
* Add context_mapping_service, which will use these configurations to extract the context from the graph and return the results as json.

This PR also defines generalized helper methods for fetching configs from the config hash.  This includes fetching required configs, boolean configs, and configs values that need to be converted to a symbol.

cam156 commented on app/models/qa/iri_template/variable_map.rb:
```diff
@@ -13,23 +13,23 @@ class VariableMap
+      def initialize(variable_map)
+        @variable = Qa::LinkedData::Config::Helper.fetch_required(variable_map, :variable, nil)
+        @required = Qa::LinkedData::Config::Helper.fetch_boolean(variable_map, :required, nil)
+        @default = variable_map.fetch(:default, '').to_s
```
@elrayle No need for changes, but I am noticing an inconstancy here.  In Some cases you use the Helper class and in other cases you access the map directly.  It might be better in the long run to be consistent.

elrayle replied:
This makes sense.  I'll make that change.

cam156 commented on app/models/qa/iri_template/variable_map.rb:
```diff
@@ -13,23 +13,23 @@ class VariableMap
       # TODO: When implementing more complex query substitution, simple_value is used when template url specifies variable as {var_name}.
       # Value to use in substitution, using default if one isn't passed in
-      # @param [Object] value to use if it exists
+      # rparam [Object] value to use if it exists
```
@elrayle this looks like a typo

elrayle replied:
nice catch

cam156 commented on app/models/qa/linked_data/config/context_map.rb:
```diff
@@ -0,0 +1,75 @@
+        attr_reader :context_map, :groups
+        private :context_map, :groups
+
+        # @param [Hash] context_map that defines groups and properties for additional context to display during the selection process
```
👍 Comments look great!

cam156 commented on app/models/qa/linked_data/config/context_map.rb:
```diff
@@ -0,0 +1,75 @@
+          extract_properties
+        end
+
+        def group_label(group_id)
```
You are allowing the user to get a list of properties back, but not a list of groups.  They need to know the id to get at group information.  I'm guessing this is how you are using the data, but getting a list of groups seems like it might be nice.

elrayle replied:
I am trying to only surface the public API what is actually being used.  When started out, I was surfacing what I thought would be useful and decided against that to avoid public API bloat.  If someone requests the list of groups, it can easily be added at that time.

elrayle commented:
* fixed typo rparam to @param
* fixed typo @returns to @return
* consistently use helper class for all access of the config hash (in new and recently refactored classes only)
